src/LibraryManager.Vsix/UI/InstallDialogTestContract.cs

```diff
@@ -0,0 +1,14 @@
+﻿using System.Threading.Tasks;
+
+namespace Microsoft.Web.LibraryManager.Vsix.UI
+{
+    /// <summary>
+    /// This class gives apex access to opened - add client side libraries dialog.
+    /// </summary>
+    internal class InstallDialogTestContract
+    {
+        public static IInstallDialogTestContract Window;
+        // The TaskCompletionSource lets apex know when the dialog is open.
+        public static TaskCompletionSource<bool> WindowIsUp = new TaskCompletionSource<bool>();
```

Comment:
>TaskCompletionSource [](start = 22, length = 20)

Something else bugging me is actually having the task (or EventWaitHandle) on this side, it doesn't seem like the product should know that someone needs to wait for this, it should just provide the data. What do you think about something like the following?

*** rename IInstallDialogTestContract => IInstallDialog 

*** change InstallDialogTestContract to the following
internal class InstallDialogProvider
{
    private static IInstallDialog _installDialog;
    public static IInstallDialog InstallDialog
    {
        get => _installDialog;
        set {
           _installDialog= value;
           InstallDialogChanged(...);
        }
    }

    public static event InstallDialogChanged;
}

*** change InstallDialog to call
InstallDialogProvider.Window = ...

*** And InstallDialogTestService would have the understanding to use tasks (it's got the complicated task stuff)
TaskCompletionSource<bool> _windowChangedTask = new TaskCompletionSource<bool>();

InstallDialogProvider.WindowChanged += OnWindowChanged;
if (InstallDialogProvider.Window == null)
{
    await _windowChangedTask;
}
InstallDialogProvider.WindowChanged -= OnWindowChanged;

private void OnWindowChanged(...)
{
    _windowChangedTask.SetResult(true);
}

Reply:
Made some changes in iteration 5. Let me know what you think.

Thanks!

---
In reply to: [205942824](https://github.com/aspnet/LibraryManager/pull/332#discussion_r205942824) [](ancestors = 205942824)